smithy-model/src/main/java/software/amazon/smithy/model/neighbor/NodeQuery.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.model.neighbor;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.utils.ListUtils;
+
+/**
+ * Searches {@link Node}s to find matching children. Each search
+ * condition is executed on the result of the previous search,
+ * and the results are aggregated.
+ */
+final class NodeQuery {
+    private final List<Query> queries = new ArrayList<>();
+
+    NodeQuery() {
+    }
+
+    NodeQuery self() {
+        queries.add(Stream::of);
+        return this;
+    }
+
+    NodeQuery member(String name) {
+        queries.add(node -> {
+            if (node == null || !node.isObjectNode()) {
+                return Stream.empty();
+            }
+            return node.expectObjectNode().getMember(name).map(Stream::of).orElse(Stream.empty());
+        });
+        return this;
+    }
+
+    NodeQuery anyMember() {
```

Comment:
seems like several of these could be private static final instances of a query rather than recreated each time.